lang/java/avro/src/main/java/org/apache/avro/generic/GenericDatumReader.java

```diff
@@ -265,7 +267,17 @@ protected void readField(Object record, Field field, Object oldDatum, ResolvingD
    * representations. By default, returns a GenericEnumSymbol.
    */
   protected Object readEnum(Schema expected, Decoder in) throws IOException {
-    return createEnum(expected.getEnumSymbols().get(in.readEnum()), expected);
+    List<String> enumSymbols = expected.getEnumSymbols();
+    int ordinal = in.readEnum();
+    if (ordinal >= enumSymbols.size()) {
+      if (expected.getEnumDefault() == null) {
+        throw new AvroTypeException("Unknown Enum Ordinal " + ordinal);
+      }
+
+      return createEnum(expected.getEnumDefault(), expected);
```

Comment:
that contradicts the documentation:
if both are enums: if the writer’s symbol is not present in the reader’s enum and the reader has a default value, then that value is used, otherwise an error is signalled.

If the default value is specified - it should be used, otherwise - throw an exception. Absen of this rule, it will be hard to promote schema evolution and all clients will need to be aware of the current writer schema before the writer could use it. Which means in reality, only one schema can be used and no evolution of enums suported

Reply:
In this case, serialized data wasn't produced with a known writer schema and the deserializer should use a default value if present in the reader schema.

Reply:
Without a (correct!) writer schema, Avro data is basically garbage. Attempting to fix this is a laudible effort, but also introduces bugs.

What happens, for example if data was written with `{"type":"enum","name":"Switch","symbols":["OFF","DIMMED1","DIMMED2","DIMMED3","FULL"]}`, but subsequently read with the older `{"type":"enum","name":"Switch","symbols":["ON","OFF"]}`?

IMHO, throwing an exception is the correct way to handle this.

Reply:
The requirement to have the correct writer's schema during the read operation makes Avro unusable in real-world production scenarios. 
If this requirement is enforced then schemas can't be updated ever since it's impossible to have all message consumers switch to a new schema at the same time as the writer without downtime.
Developers are in charge of schema evolution and, if stated properly in documentation, situations like change of enum order can be avoided

Reply:
On the contrary: schema evolution is handled by schema resolution. This is the process that translates data written with the "write schema" into a structure defined by the "read schema" **while reading the data**.

See https://avro.apache.org/docs/1.11.1/specification/#schema-resolution

All consumers read the data into a structure defined by the read schema they were compiled/deployed with. This is static and changes only when the consumers are ready.

To do so, consumers must have access to the write schema used to write the data, for example by reading the schema embedded in a `.avro` file, or via a service implemented by the `SchemaStore` interface. Schema resolution/evolution works as long as both schemas are compatible.

Reply:
Nothing in the doc says the reader should have access to writer's schema. Contrary, it says:
```
the reader may be programmed to read data into a different schema. For example, if the data was written with a different version of the software than it is read
```
If we insist on the reader to always has the writes' schema, it will mean that schema evolution is not possible at all in real-world scenarios.
There might be several schema versions, and, as long as they are backward compatible (nullable fields, fields are added) reader should be able to read the data.

Reply:
In "Getting started (Java)", the doc says:
> The writer’s schema is needed to know the order in which fields were written, while the reader’s schema is needed to know what fields are expected and how to fill in default values for fields added since the file was written. If there are differences between the two schemas, they are resolved according to the Schema Resolution specification.

More on schema evolution has been written, for example by Martin Kleppmann: https://martin.kleppmann.com/2012/12/05/schema-evolution-in-avro-protocol-buffers-thrift.html  
And on stackoverflow:  
https://stackoverflow.com/questions/73323112/does-avro-supports-schema-evolution/73324135  
https://stackoverflow.com/questions/15350509/avro-schema-evolution

There is more to be found, but I'll leave that as an exercise for the reader.